### PR TITLE
Integrate sequence-packing scheduler into training loop and wire varlen dataset into GPT pretraining

### PR DESCRIPTION
**Pull request by ilml**
## What does this PR do?

Integrates the sequence-packing scheduler into `train_step`/`evaluate` and wires the varlen dataset into the pretraining CLI, dataloader, and dataset provider (`--use-varlen-dataset`). Final PR of the [#3386](https://github.com/NVIDIA/Megatron-LM/pull/3386) split series tracked in the [Dynamic Context Parallelism project](https://github.com/orgs/NVIDIA/projects/297) — after this, the full THD/sequence-packing E2E from #3386 is landed. Original changes by @xiaoyao0115 in #3386.

**Series and merge order** (reviews can run in parallel):

| PR | Content | Merge order |
|---|---|---|
| #5902 | config field, CLI arg, validation | 1 (independent) |
| #6684 | core packing scheduler | 1 (independent) |
| #6685 | mock SFT + varlen datasets | 2 |
| **this PR** | scheduler + varlen wiring into training loop | 3 — last |

> **Stacking note:** base is `pull-request/6685` so Files-changed shows only this PR's two layers (training-loop integration, varlen wiring). It will be retargeted to `main` as soon as #6685 is approved, before any merge.

## Contribution process

- [x] Draft PR per contributing guidelines
- [x] Commits signed off (DCO)

### CODE REVIEW

**Comment by Phlip79** [comment]
What about for `HybridModel`?

File: megatron/training/datasets/data_samplers.py [hunk 1]
```diff
@@ -45,14 +45,16 @@ def build_pretraining_data_loader(dataset, consumed_samples):
...
-
     if split == Split.valid and args.full_validation:
         batch_sampler = MegatronFullValidationSampler(
             total_samples=len(dataset),
             data_parallel_rank=mpu.get_data_parallel_rank(),
             data_parallel_size=mpu.get_data_parallel_world_size())
     elif args.dataloader_type == 'single':
-        if args.hybrid_context_parallel:
+        if (
+            getattr(args, "hybrid_context_parallel", False)
```
Comment (jaredcasper) [inline megatron/training/datasets/data_samplers.py:55]: why the switch to getattr?

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -3777,6 +3806,47 @@ def _add_sft_args(parser):
     )
     return parser
 
+
+def _add_varlen_dataset_args(parser):
+    group = parser.add_argument_group(title='varlen dataset')
+    group.add_argument(
```
Comment (jaredcasper) [inline megatron/training/arguments.py:3812]: Please match formatting with the rest of the file.

File: megatron/training/training.py [hunk 3]
```diff
@@ -710,6 +694,24 @@ def print_datetime(string, override_timestamp=None):
         time_str = datetime.fromtimestamp(override_timestamp).strftime('%Y-%m-%d %H:%M:%S.%f')
     print_rank_0(f'[{string}] datetime: {time_str} ')
 
+# Per-iteration packed-sequence (THD) accumulator. The tensor holds TWO stats,
```
Comment (jaredcasper) [inline megatron/training/training.py:697]: Why move these down here?

File: megatron/training/arguments.py [hunk 4]
```diff
@@ -1652,6 +1653,34 @@ def validate_args(args, defaults={}):
     if args.ckpt_format == "fsdp_dtensor":
         assert args.use_megatron_fsdp, "--ckpt-format fsdp_dtensor is only tested with Megatron FSDP."
 
+    # --use-varlen-dataset: independent of --sft. Cannot be combined with --sft
+    # because they are mutually-exclusive top-level dataset selectors that both
+    # drive the packed-sequence (THD) path.
+    if args.use_varlen_dataset:
```
Comment (jaredcasper) [inline megatron/training/arguments.py:1659]: Can these go in a dataclass's post_init?
